src/core/qgsfield.cpp

```diff
@@ -589,6 +589,29 @@ bool QgsField::convertCompatible( QVariant &v, QString *errorMessage ) const
     return false;
   }
 
+  // Handle referenced geometries (e.g. from additional geometry fields)
+  if ( d->type == QVariant::String && v.type() == QVariant::UserType && v.canConvert<QgsReferencedGeometry>( ) )
+  {
+    const QgsReferencedGeometry geom { v.value<QgsReferencedGeometry>( ) };
+    if ( geom.isNull() )
+    {
+      v = QVariant( d->type );
+    }
+    else
+    {
+      // EWKT only if authid is EPSG, not sure if it makes sense otherwise
+      if ( geom.crs().authid().startsWith( QStringLiteral( "EPSG:" ) ) )
```

Comment:
We should move this to QgsReferencedGeometry. Maybe something like:

    /**
      * Returns the geometry converted to a EWKT string.
      *
      * \param precision decimal precision to use for WKT coordinates
      * \param ok TRUE if geometry can be represented as a EWKT string, FALSE otherwise
      * \returns geometry as a EWKT string.
    */
    QString asEWkt( int precision = 17, bool* ok = nullptr SIP_OUT ) const;

Reply:
@nyalldawson  I've removed the E from WKT as we discussed, I guess there is no need for this method anymore.